src/main/java/com/datadoghq/datadog_lambda_java/DDLambda.java

```diff
@@ -103,10 +111,69 @@ public DDLambda(Headerable req, Context cxt) {
         this.tracing = new Tracing(req);
         this.tracing.submitSegment();
         addTraceContextToMDC();
-        addTagsToSpan(cxt);
+        startSpan(req.getHeaders(), cxt);
     }
 
-    private void addTagsToSpan(Context cxt) {
+    /**
+     * startSpan is called by the DDLambda constructors. If there is a dd-agent-java present, it will start
+     * a span. If not, startSpan is a noop.
+     * @param headers are the headers from the Lambda request. If Headers are null or empty, distributed tracing
+     *                is impossible but a span will still start.
+     * @param cxt is the Lambda Context passed to the Handler function.
+     */
+    private void startSpan(Map<String,String> headers, Context cxt){
+        //If the user has not set DD_TRACE_ENABLED=true, don't start a span
+        if(!checkTraceEnabled()){
+            return;
+        }
+
+        String functionName = "";
+        if (cxt != null){
+            functionName = cxt.getFunctionName();
+        }
+
+        //Get the Datadog tracer, if it exists
+        Tracer tracer = GlobalTracer.get();
+        //Datadog tracer will be able to extract datadog trace headers from an incoming request
+        SpanContext parentContext = tracer.extract(Builtin.HTTP_HEADERS, new TextMapAdapter(headers));
+
+        SpanBuilder spanBuilder = tracer.buildSpan(functionName).asChildOf(parentContext);
+        spanBuilder = addDDTags(spanBuilder, cxt);
+        Span thisSpan = spanBuilder.start();
+
+        //Hang on to the scope, we'll need it later to close.
+        this.tracingScope = tracer.activateSpan(thisSpan);
+    }
+
+    /**
+     * Finish the active span. If you have installed the dd-trace-java Lambda
+     * layer, you MUST call DDLambda.finish() at the end of your Handler
+     * in order to finish spans.
```

Comment:
In the other libraries, we are able to run code automatically at the end of the function execution the finish the span (and do other stuff, like send enhanced metrics). Is it not possible to do that in Java as well? I think requiring customers to do it manually could potentially lead to confusion.

Reply:
That's on the table for future improvements, but it's difficult to implement right now. Much more so than in interpreted languages. One strategy we're looking into is potentially porting this logic into dd-agent-java (since that is already doing a lot of reflective stuff to decorate other classes), but we have some concerns about being beholden to dd-agent-java's release cycle. TBD.

Reply:
I do agree that ideally we would have an @Decorator added to a handler function, but I think this compromise is okay for now.